src/keyring/KeyringController.ts

```diff
@@ -730,6 +741,114 @@ export class KeyringController extends BaseController<
     await this.#keyring.persistAllKeyrings();
     await this.fullUpdate();
   }
+
+  // Ledger Related methods
+
+  private async addLedgerKeyring(): Promise<LedgerKeyring> {
+    return await this.#keyring.addNewKeyring(KeyringTypes.ledger);
+  }
+
+  /**
+   * Retrieve the existing LedgerKeyring or create a new one.
+   *
+   * @returns The stored Ledger Keyring
+   */
+  async getLedgerKeyring(): Promise<LedgerKeyring> {
+    const keyring = this.#keyring.getKeyringsByType(KeyringTypes.ledger)[0];
+
+    if (keyring) {
+      return keyring;
+    }
+
+    return await this.addLedgerKeyring();
+  }
+
+  /**
+   * Connects to the ledger device by requesting some metadata from it.
+   *
+   * @param transport - The transport to use to connect to the device
+   * @param deviceId - The device ID to connect to
+   * @returns The name of the currently open application on the device
+   */
+  async connectLedgerHardware(
+    transport: Transport,
+    deviceId: string,
+  ): Promise<string> {
+    const keyring = await this.getLedgerKeyring();
+    keyring.setTransport(transport, deviceId);
+
+    const { appName } = await keyring.getAppAndVersion();
+
+    return appName;
+  }
+
+  /**
+   * Retrieve the first account from the Ledger device.
+   *
+   * @returns The default (first) account on the device
+   */
+  async unlockLedgerDefaultAccount(): Promise<{
+    address: string;
+    balance: string;
+  }> {
+    const keyring = await this.getLedgerKeyring();
+    const oldAccounts = await this.#keyring.getAccounts();
+    await this.#keyring.addNewAccount(keyring);
+    const newAccounts = await this.#keyring.getAccounts();
+
+    this.updateIdentities(newAccounts);
+    newAccounts.forEach((address: string) => {
+      if (!oldAccounts.includes(address)) {
+        if (this.setAccountLabel) {
+          this.setAccountLabel(address, `${keyring.getName()} 1`);
+        }
+        this.setSelectedAddress(address);
+      }
+    });
+    await this.#keyring.persistAllKeyrings();
+    this.fullUpdate();
+
+    const address = await keyring.getDefaultAccount();
+    return {
+      address,
+      balance: `0x0`,
+    };
+  }
+
+  /**
+   * Automatically opens the Ethereum app on the Ledger device.
+   */
+  async openEthereumApp(): Promise<void> {
```

Comment:
Since all of these methods are public in the KeyringController interface, what do you think about putting "Ledger" somewhere in the names of this method and `closeRunningApp` just to communicate that these are Ledger-only? Say, `openEthereumAppOnLedger` and `closeRunningAppOnLedger`?

Reply:
Also, I don't see tests for either of these methods, should they be added?

Reply:
[967d76b](https://github.com/MetaMask/controllers/pull/862/commits/967d76b01da77b7707a270eea022562b057389d4)

Reply:
Hmm you're right, I'm not sure how to test those methods from the top of my head since they're from the `LedgerKeyring`. If you think it's important to add those tests I can make time to do it, if not I can address them in a future PR

Reply:
What do you think about stubbing out the keyring object (i.e. the result of `getLedgerKeyring`) and test that the right methods get called in these cases?

Reply:
I'm not sure how much value it would add @mcmire, since everything will be a mock. Do you think it's worth to spend time doing it?

Reply:
After writing the tests for the rest I kind of realised they don't add much value either as pretty much everything is mocked out and proves really nothing. Doesn't even prove the methods are really there. If you need to increase code coverage or you have a strict policy about it, then you could do it however I really don't think you'll get any extra safety / benefits just add to the test execution time.

Reply:
I agree that stubbing out the Ledger keyring class is a reasonable way to approach testing this. You can [mock the entire keyring module using Jest](https://jestjs.io/docs/es6-class-mocks#the-4-ways-to-create-an-es6-class-mock). Another option would be to setup a mock transport.

In either case, it is not true that the test "doesn't even prove the methods are really there", as Jest won't create stubs for methods that don't exist. Even if we're manually mocking it, we can do so in a way that ensures the test blows up if it tries to stub out a non-existent method.

The value would be that it ensures this controller is using the keyring in the correct way, and that the correct internal actions are performed. None of the code directly under test in this repo needs to be mocked.

Reply:
Thanks everyone for your comments, I'll update the PR appropriately

Reply:
@mcmire @Gudahtt @zoltana-ledger let me know what you think 

[ac35775](https://github.com/MetaMask/controllers/pull/862/commits/ac35775f615b0b519f21f4dfb40fcd0317ddec1c)

Reply:
Oops, forgot to follow up, this looks good to me.